pkg/controller/nodeipam/ipam/multi_cidr_priority_queue.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ipam
+
+import (
+	"math"
+
+	cidrset "k8s.io/kubernetes/pkg/controller/nodeipam/ipam/multicidrset"
+)
+
+// A PriorityQueue implementation based on https://pkg.go.dev/container/heap#example-package-PriorityQueue
+
+// An PriorityQueueItem is something we manage in a priority queue.
+type PriorityQueueItem struct {
+	clusterCIDR *cidrset.ClusterCIDR
+	// labelMatchCount is the first determinant of priority.
+	labelMatchCount int
+	// selectorString is a string representation of the labelSelector associated with the cidrSet.
+	selectorString string
+	// index is needed by update and is maintained by the heap.Interface methods.
+	index int // The index of the item in the heap.
```

Comment:
this index seems a leftover of the example in https://pkg.go.dev/container/heap#example-package-PriorityQueue . at least I can't see its

Reply:
It's required for the heap implementation, however we might shun away from using the heap altogether, might need to handle this in another PR post alpha though

Reply:
ahh good to know

Reply:
ok, I see, it is needed if you want to modify items that are already in the queue ...